File: fundamental_algorithms/heap_sort.cpp
#include <stdlib.h>
#include <unistd.h>
#include <iostream>

#include <vector>

static void heapify(std::vector<int> &array, size_t idx, size_t len)
{
    size_t left =  (idx << 1) + 1;
    size_t right = (idx << 1) + 2;

    std::cout << idx << ",";

    size_t largest = idx;
    if (left < len && array[left] > array[idx])
        largest = left;
    if (right < len && array[right] > array[largest])
        largest = right;

    if (largest != idx) {
        int tmp = array[idx];
        array[idx] = array[largest];
        array[largest] = tmp;

        heapify(array, largest, len);
    } else {
        std::cout << "ok" << std::endl;
    }
}

static void build_heap(std::vector<int> &array)
{
    for (int i = array.size() / 2 - 1; i >= 0; --i) {
        heapify(array, i, array.size());
    }
}

void heap_sort(std::vector<int> &array)
{
    build_heap(array);
    for (int i = array.size() - 1; i > 0; --i) {
        int tmp = array[0];
        array[0] = array[i];
        array[i] = tmp;
        heapify(array, 0, i);
    }
}

int main(int argc, char *argv[])
{
    std::vector<int> array;
    for (int i = 0; i < 10; ++i)
        array.push_back(i);

    heap_sort(array);
    std::cout << std::endl;

    for (size_t i = 0; i < array.size(); ++i)
        std::cout << array[i] << " ";
    std::cout << std::endl;
    return 0;
}

File: fundamental_algorithms/heap_sort.py
def heapify(L, idx, size):
    l = idx * 2 + 1
    r = idx * 2 + 2

    largest = idx
    if l < size and L[l] < L[idx]:
        largest = l
    if r < size and L[r] < L[largest]:
        largest = r

    if largest is not idx:
        t = L[idx]
        L[idx] = L[largest]
        L[largest] = t
        heapify(L, largest, size)


def build_heap(L):
    for i in range(len(L) / 2 - 1, -1, -1):
        heapify(L, i, len(L))


def heap_sort(L):
    build_heap(L)
    for i in range(len(L) - 1, 0, -1):
        t = L[0]
        L[0] = L[i]
        L[i] = t
        heapify(L, 0, i)


def heap_insert(L, x):
    pass


def heap_maximun(L):
    pass


def extract_max(L):
    pass


def increase_key(L, idx, x):
    '''中文注释'''
    pass


def quick_sort(L, low, high):
    l = low
    #    l = high / 2
    h = high
    key = L[l]

    if (l >= h):
        return

    while l < h:
        while l < h and L[h] > key:
            h = h - 1
        L[l] = L[h]
        while l < h and L[l] < key:
            l = l + 1
        L[h] = L[l]
    L[l] = key

    quick_sort(L, low, l - 1)
    quick_sort(L, h + 1, high)


def count_sort(L, low, high):
    if low >= high:
        return

    count = []
    for i in range(high - low + 2):
        count.append(0)
    for item in L:
        if item > high:
            continue
        count[item - low] = count[item - low] + 1

    holder = 0
    for i in range(len(count)):
        for j in range(count[i]):
            L[holder] = i + low
            holder = holder + 1


L = [1, 6, 7, 9, 2, 3, 5, 4, 8, 10]
print '-->'
# print L
# heap_sort(L)
# print L

print 'quick_sort'
quick_sort(L, 0, len(L) - 1)
# count_sort(L, 1, 10)
print L


L = [1, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5]
single = 0
for l in L:
    single = single ^ l
print single
